tests/reconfiguration.py

```diff
@@ -303,6 +305,45 @@ def test_join_straddling_primary_replacement(network, args):
     return network
 
 
+def test_retiring_nodes_emit_at_most_one_signature(network, args):
+    primary, _ = network.find_primary()
+
+    # Force ledger flush of all transactions so far
+    network.get_latest_ledger_public_state()
+    ledger = ccf.ledger.Ledger(primary.remote.ledger_paths())
+
+    NODES = "public:ccf.gov.nodes.info"
+    SIGS = "public:ccf.internal.signatures"
+
+    retiring_nodes = set()
+    retired_nodes = set()
+    for chunk in ledger:
+        for tr in chunk:
+            tables = tr.get_public_domain().get_tables()
+            if NODES in tables:
+                nodes = tables[NODES]
+                for nid, info_ in nodes.items():
+                    info = json.loads(info_)
+                    if info["status"] == "Retired":
+                        retiring_nodes.add(nid)
+
+            if SIGS in tables:
+                sigs = tables[SIGS]
+                assert len(sigs) == 1, sigs.keys()
+                (sig_,) = sigs.values()
+                sig = json.loads(sig_)
+                assert (
+                    sig["node"] not in retired_nodes
+                ), f"Unexpected signature from {sig['node']}"
+                retired_nodes |= retiring_nodes
+                retiring_nodes = set()
+
+    assert not retiring_nodes, (retiring_nodes, retired_nodes)
+    LOG.info("{} nodes retired throughout test", len(retired_nodes))
```

Comment:
Should this be an invariant that we always check for in `ledger.py`? The `LedgerValidator` already keeps track of the nodes' statuses as it iterates over ledger transactions so we could easily check that here.

Reply:
I don't think so, I think it will cause problems for ledgers that have a prefix created by an older node.

Reply:
I'm not sure I follow. The identity of the retired primary will be present in the ledger up to the next signature after its retirement, which should be fine?

Reply:
But there will often be more than one post-retirement signature?

Reply:
Only _one_ signature emitted by the _just-retired_ primary though, as this PR implements?